CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
+
+This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.
+
+To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
+```
+PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`
+GetPlumoProofMsg - Request plumo proof(s) from a peer
+```
+`GetPlumoProofMsg` will be implemented by having the requester send a set of proof IDs that they hold, and ask for any they are missing. This may be optimized using a bloom filter as well.
+
+The `les` protocol must then also be extended via at least two messages:
+```
+GetEpochsMsg - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
+EpochsMsg - Receives a mix of header(s) and proof(s) from the server.
+```
+
+The `les` client must then be modified to send a `GetEpochsMsg` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.
+
+Upon receiving the `EpochsMsg` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
```

Comment:
Could you elaborate on this a bit? I don't quite understand from this description.

Reply:
This is an area I don't have great context on, I was tipped off that this will likely be an issue by asa.

My understanding is that the underlying `lightchain` data is likely not readily accessible for the light client when syncing in `downloader` so there will need to be changes to expose this information and may require new methods to update the validator set

Reply:
I think what you might be talking about is that we need some way to be able to write an Istanbul validator snapshot for a given epoch after a proof is verified. Currently, I believe the only way for light clients to write a snapshot is to validate an epoch header.

Reply:
Yup exactly, that's my understanding of the necessary changes